### PR TITLE
feat: enable resend email menu

### LINKED ISSUE #473: Resending Emails to Recipients
# Document signing "Resend" feature

This feature should allow document owners to send reminders to one or more document recipients. It should be a modal popup that looks like this:

Screenshot 2023-10-02 at 15 43 59

The user should be able to only send a reminder to the recipients who didn't sign the document. The recipients that signed the document could have a green tick instead of the checkbox, or they could be hidden altogether.

The current "Resend" button can be found on the /documents page.

Screenshot 2023-10-02 at 15 45 36

Feel free to ask any questions.

### PR DESCRIPTION
**Pull request by G3root**
fixes #473

### CODE REVIEW

**Comment by dguyen** [comment]
Hey @G3root

Sorry about this but we had a spec for this feature which wasn't uploaded as part of the issue. It has now been updated if you could please have a look:

https://github.com/documenso/documenso/issues/473

Thanks for all your contributions so far!

**Comment by G3root** [comment]
@dguyen pushed some changes


[Screencast 2023-10-03 23:26:15.webm](https://github.com/documenso/documenso/assets/84864519/e772bfcc-b5f2-4049-b763-d46072ee7e8e)

File: packages/lib/server-only/document/send-document.tsx [hunk 1]
```diff
@@ -11,9 +11,10 @@ import { DocumentStatus, SendStatus } from '@documenso/prisma/client';
 export type SendDocumentOptions = {
   documentId: number;
   userId: number;
+  resendEmails?: string[];
 };
 
-export const sendDocument = async ({ documentId, userId }: SendDocumentOptions) => {
+export const sendDocument = async ({ documentId, userId, resendEmails }: SendDocumentOptions) => {
```
Comment (dephraiim) [inline packages/lib/server-only/document/send-document.tsx:17]: Can you avoid this and instead create a new router for resending a document?

You can also modify the subject of the emails that you send.

Comment (dephraiim) [inline packages/lib/server-only/document/send-document.tsx:17]: Can you avoid this and instead create a new router for resending a document?

You can also modify the subject of the emails that you send.

**Review by dephraiim — CHANGES_REQUESTED** [review]

**Comment by G3root** [comment]
@dephraiim pushed some changes

File: packages/lib/server-only/document/send-document.tsx [hunk 2]
```diff
@@ -8,12 +8,31 @@ import { renderCustomEmailTemplate } from '@documenso/lib/utils/render-custom-em
...
 
-export const sendDocument = async ({ documentId, userId }: SendDocumentOptions) => {
+type CustomEmail = {
+  email?: {
+    subject: string;
+    message: string;
+  };
+};
+
+export type resendDocumentOptions = {
```
Comment (dephraiim) [inline packages/lib/server-only/document/send-document.tsx:25]: Would prefer `ResendDocumentOptions`

File: packages/lib/server-only/document/send-document.tsx [hunk 3]
```diff
@@ -8,12 +8,31 @@ import { renderCustomEmailTemplate } from '@documenso/lib/utils/render-custom-em
...
+  resendEmails: string[];
+} & SendDocumentOptions &
+  CustomEmail;
+
+type getSendableDocumentOptions = {
```
Comment (dephraiim) [inline packages/lib/server-only/document/send-document.tsx:30]: Would prefer `GetSendableDocumentOptions`

File: packages/lib/server-only/document/send-document.tsx [hunk 4]
```diff
@@ -44,7 +61,26 @@ export const sendDocument = async ({ documentId, userId }: SendDocumentOptions)
...
+  return { document, user, customEmail };
+}
+
+type TGetSendableDocument = Awaited<ReturnType<typeof getSendableDocument>>;
 
+type SendDocumentsOptions = {
+  resendEmails?: string[];
+} & TGetSendableDocument;
+
+async function mailDocuments({ customEmail, document, resendEmails, user }: SendDocumentsOptions) {
```
Comment (dephraiim) [inline packages/lib/server-only/document/send-document.tsx:83]: `mailDocuments` should not have `SendDocumentsOptions` a prop type

**Review by dephraiim — CHANGES_REQUESTED** [review]

**Comment by Mythie** [comment]
Guess who's back... it's me here to get this bad boy over the line 🤝🏻

**Comment by adithyaakrishna** [comment]
based

**Review by Mythie — APPROVED** [review]

**Comment by Mythie** [comment]
We did it 💪 Thanks Nafees!

thank-you-stronk

**Comment by G3root** [comment]
> We did it 💪 Thanks Nafees!
>

**Pull request merged**